pandas/core/groupby/groupby.py

```diff
@@ -969,22 +969,17 @@ def reset_identity(values):
             result = concat(values, axis=self.axis)
             ax = self._selected_obj._get_axis(self.axis)
 
-            if isinstance(result, Series):
-                result = result.reindex(ax)
+            # this is a very unfortunate situation
+            # we can't use reindex to restore the original order
+            # when the ax has duplicates
+            # so we resort to this
+            # GH 14776, 30667
+            if ax.has_duplicates:
+                indexer, _ = result.index.get_indexer_non_unique(ax.values)
+                indexer = algorithms.unique1d(indexer)
```

Comment:
i don't believe you actually need to do this which defeats the purpose of non-unique indexing

Reply:
I think maybe we are thinking of different things? 
Here because we have `mutated == False` (more on that later), we want to restore the original order by index.
The problem is, when index is duplicated, you can't do `reindex` but resort to `Index.get_indexer_non_unique`, which produces duplicated results, see below.
```python
>>> import pandas as pd
>>> pd.__version__
'0.26.0.dev0+1724.gc5b793a2d'

>>> index = pd.Index([2, 1, 2])
>>> ser = pd.Series([0, 1, 2], index=index)
>>> ser
2    0
1    1
2    2
dtype: int64

>>> indexer, _ = index.get_indexer_non_unique([1, 2, 2])
>>> indexer
array([1, 0, 2, 0, 2], dtype=int64)

>>> ser.take(indexer)  # This doesn't give us the reordered result
1    1
2    0
2    2
2    0
2    2
dtype: int64

>>> from pandas.core import algorithms
>>> indexer = algorithms.unique1d(indexer)
>>> ser.take(indexer)  # This does give us what we want
1    1
2    0
2    2
dtype: int64

>>> ser.reindex([1, 2, 2])  # This results in ValueError
Traceback (most recent call last):
  ...
ValueError: cannot reindex from a duplicate axis
```

Perhaps are you thinking of the refactoring mentioned in #14927?
If we can actually deprecate the behavior of trying to guess if an object is mutated and just return a multi-indexed `DataFrame` like `groupby.transform` does, then it is true we don't need to deduplicate here. What I wrote here is more like a smaller scale fix/patch.

Am I understanding this correctly? Let me know.